### PR TITLE
VITIS-6990: xrt::xclbin API to obtain interface_uuid

### PR DESCRIPTION
**Pull request by vboggara-xilinx**
#### Problem solved by the commit
Added below API for xrt::xclbin to obtain interface_uuid 
```
//get_interface_uuid() - Get the Interface UUID of the xclbin. 
//return  - Interface UUID of the xclbin

XCL_DRIVER_DLLESPEC
uuid
get_interface_uuid() const;
```
#### Bug / issue (if any) fixed, which PR introduced the bug, how it was discovered
https://jira.xilinx.com/browse/VITIS-6990
#### How problem was solved, alternative solutions (if any) and why they were rejected
1. xclbin generation : updating xclbin header content rom_uuid with interface_uuid from partition_metadata.
2. In xrt::xclbin, reading rom_uuid from axlf header and return it in new api( get_interface_uuid)
#### Risks (if any) associated the changes in the commit
Low risk as it is new API.
#### What has been tested and how, request additional testing if necessary
Tested locally some cases.
#### Documentation impact (if any)
Yes, Need to add this new API to the XRT::XCLBIN documentation.

### CODE REVIEW

**Comment by vboggara-xilinx** [comment]
seems build failed because of infra issue. please retest it.

**Review by stsoe — APPROVED** [review]
Thank you.  Much simpler.

**Review by chvamshi-xilinx — APPROVED** [review]
Looks good. Couple of things to note
1) Please file a CR on xclbinutil to merge these changes on Vitis side also
2) Update one of the existing test to exercise interface_uuid()

**Comment by vboggara-xilinx** [comment]
@rozumx , pl review xclbinutil changes.

**Comment by vboggara-xilinx** [comment]
> Looks good. Couple of things to note
> 
> 1. Please file a CR on xclbinutil to merge these changes on Vitis side also
> 2. Update one of the existing test to exercise interface_uuid()

sure @chvamshi-xilinx , will do.

**Review by rozumx — CHANGES_REQUESTED** [review]
While the code is simply, this is a big architectural and functional change.  In a nutshell, the FeatureRom functionality is obsoleted in the xclbin and xclbinutil. 

This is O.K. for the "classic" platforms are no long supported.

Being that this is case, the code needs to cleaned and refactored removing the FeatureRom functionality.  This includes:

1. Renaming `axlf_header.rom_uuid` to interface UUID.  
2. Adding a DRC check to xclbinutil to check for the existence of 1 interface UUID in the partition_metadata (if existed).  Anything else will produce an error.
3. Remove `axlf_header.m_platformId` and `axlf_header.m_featureId`.  After removing it the union can be removed resulting in the  `axlf_header.interface_uuid`.
4. Remove key support for setting "FeatureRomTimestamp" and rename the existing entry to `m_reserved1`.  See XclBinClass::setKeyValue()

**Comment by chvamshi-xilinx** [comment]
> 

Thanks @rozumx  for the review. I have a question on #2 point. ULPs have only one interface uuid, but
there would be two interface uuids (resolved and exposed) for the PLP xsabins. 
xclbinutil gets used for generating platform blp/plp/ulp xsabin/xclbin. Adding an error may cause problem for PLP xsabin generation. 
Let me know if am missing anything?

**Review by uday610 — APPROVED** [review]

File: src/runtime_src/tools/xclbinutil/XclBinClass.cxx [hunk 1]
```diff
@@ -1234,6 +1229,27 @@ XclBin::appendSections(ParameterSectionData& _PSD)
     pSection->purgeBuffers();
     pSection->readJSONSectionImage(ptPayload);
 
+    if (eKind == PARTITION_METADATA) {
```
Comment (rozumx) [inline src/runtime_src/tools/xclbinutil/XclBinClass.cxx:1232]: This functional should not be done in the append section.  Instead, this functionality should be done in XclBInUtilMain::main_() right before "dumping the sections" (at around line 536).

**Review by rozumx — CHANGES_REQUESTED** [review]

**Comment by chvamshi-xilinx** [comment]
> @chvamshi-xilinx Good catch. I don't have the partition_metadata before me, but if I recall correctly there isn't a way to determine which is the exposed and which is resolved. So this DRC check (item #2 mentioned above) needs to be expanded and should examine the a`xlf.m_mode` value to determine if the `axlf_header.m_interface_uuid` should be set (as well as the DRC check).

Thanks @rozumx . This is good.
@vboggara-xilinx , Please try installing blp/plp packages also with your changes sothat we dont break anything in that flow.

File: src/runtime_src/tools/xclbinutil/XclBinClass.cxx [hunk 2]
```diff
@@ -1234,6 +1229,27 @@ XclBin::appendSections(ParameterSectionData& _PSD)
...
+        // Look for the "interfaces" node
+        boost::property_tree::ptree ptInterfaces = ptPartitionMetadata.get_child("interfaces");
+        if (ptInterfaces.empty()) {
+            continue;
+        }
+        
+        // Updated header rom_uuid with interface_uuid from partition_metadata
+        for (const auto& kv : ptInterfaces) {
+            boost::property_tree::ptree ptInterface = kv.second;
+            auto sInterfaceUUID = ptInterface.get<std::string>("interface_uuid", "00000000000000000000000000000000");
```
[the remark this answers is not in the document]
Comment (vboggara-xilinx) [reply src/runtime_src/tools/xclbinutil/XclBinClass.cxx]: @rozumx , addressed all your comments and submitted changes in latest PR. please review.

File: src/runtime_src/tools/xclbinutil/XclBinClass.cxx (hunk 1 above)
Comment (vboggara-xilinx) [reply src/runtime_src/tools/xclbinutil/XclBinClass.cxx]: addressed.

File: src/runtime_src/tools/xclbinutil/XclBinClass.cxx (hunk 2 above)
[the remark this answers is not in the document]
Comment (vboggara-xilinx) [reply src/runtime_src/tools/xclbinutil/XclBinClass.cxx]: @rozumx , FeatureRomTimestamp is used in the drivers codebase. seeing some errors in driver compilation with latest change.

[the remark this answers is not in the document]
Comment (vboggara-xilinx) [reply src/runtime_src/tools/xclbinutil/XclBinClass.cxx]: @rozumx , axlf_header.FeatureRomTimestamp  is used many places in xocl driver code base, hence I am not touching this param in the code base. addressed remaining comments and submitted latest changes in the PR. please review.

**Comment by dayeh-xilinx** [comment]
retest this please.

**Review by chvamshi-xilinx — APPROVED** [review]

**Review by rozumx — CHANGES_REQUESTED** [review]
In addition, please do remove the obsoleted keys.

File: src/runtime_src/tools/xclbinutil/XclBinClass.cxx [hunk 3]
```diff
@@ -1907,8 +1902,48 @@ XclBin::addKernels(const std::string& jsonFile)
   updateKernelSections(kernels, true /*isFixedPS*/, this);
 }
 
+void
+XclBin::updateInterfaceuuid()
+{
+  XUtil::TRACE("Updating Interface uuid in xclbin");
+  // Get the PARTITION_METADATA property tree (if there is one)
+  for (auto pSection : m_sections) {
```
[the remark this answers is not in the document]
Comment (vboggara-xilinx) [reply src/runtime_src/tools/xclbinutil/XclBinClass.cxx]: fixed it.

**Comment by dayeh-xilinx** [comment]
retest this please.

**Comment by dayeh-xilinx** [comment]
retest this please.

File: src/runtime_src/tools/xclbinutil/FormattedOutput.h
Comment (rozumx) [inline src/runtime_src/tools/xclbinutil/FormattedOutput.h:1]: Copyright dates need to be updated.

File: src/runtime_src/tools/xclbinutil/FormattedOutput.cxx
Comment (rozumx) [inline src/runtime_src/tools/xclbinutil/FormattedOutput.cxx:1]: Copyright dates need to be updated

File: src/python/xclbin_binding.py
Comment (rozumx) [inline src/python/xclbin_binding.py:1]: Copyright dates need to be updated.

File: src/runtime_src/core/common/api/xrt_xclbin.cpp
Comment (rozumx) [inline src/runtime_src/core/common/api/xrt_xclbin.cpp:1]: Copyright dates need to updated.

File: src/runtime_src/core/include/experimental/xrt_xclbin.h
Comment (rozumx) [inline src/runtime_src/core/include/experimental/xrt_xclbin.h:1]: Copyright dates need to be updated.

File: src/runtime_src/tools/xclbinutil/XclBinClass.cxx
Comment (rozumx) [inline src/runtime_src/tools/xclbinutil/XclBinClass.cxx:1]: Copyright dates need to be updated.

File: src/runtime_src/tools/xclbinutil/XclBinClass.h
Comment (rozumx) [inline src/runtime_src/tools/xclbinutil/XclBinClass.h:1]: Copyright dates need to be updated.

File: tests/python/00_hello/main.py
Comment (rozumx) [inline tests/python/00_hello/main.py:1]: Copyright dates need to be updated.

File: src/runtime_src/tools/xclbinutil/FormattedOutput.h
Comment (vboggara-xilinx) [reply src/runtime_src/tools/xclbinutil/FormattedOutput.h:1]: updated

File: src/python/xclbin_binding.py
Comment (vboggara-xilinx) [reply src/python/xclbin_binding.py:1]: updated

File: src/runtime_src/tools/xclbinutil/FormattedOutput.cxx
Comment (vboggara-xilinx) [reply src/runtime_src/tools/xclbinutil/FormattedOutput.cxx:1]: updated

File: src/runtime_src/core/common/api/xrt_xclbin.cpp
Comment (vboggara-xilinx) [reply src/runtime_src/core/common/api/xrt_xclbin.cpp:1]: updated

File: src/runtime_src/core/include/experimental/xrt_xclbin.h
Comment (vboggara-xilinx) [reply src/runtime_src/core/include/experimental/xrt_xclbin.h:1]: updated

File: src/runtime_src/tools/xclbinutil/XclBinClass.cxx
Comment (vboggara-xilinx) [reply src/runtime_src/tools/xclbinutil/XclBinClass.cxx:1]: updated

File: src/runtime_src/tools/xclbinutil/XclBinClass.h
Comment (vboggara-xilinx) [reply src/runtime_src/tools/xclbinutil/XclBinClass.h:1]: updated

File: tests/python/00_hello/main.py
Comment (vboggara-xilinx) [reply tests/python/00_hello/main.py:1]: updated

**Pull request merged**